File: beecrowd/python/grafos/ex2784.py
from heapq import heappush, heappop

def dijkistra(origem, lista):
    INF = 10e10
    distancias = [INF for u in range(len(lista))]
    distancias[origem] = 0
    fila_prioridade = []
    
    heappush(fila_prioridade, (0, origem))
    
    while len(fila_prioridade) > 0:
        dist, v_atual = heappop(fila_prioridade)
        if dist > distancias[v_atual]:
            continue
        
        for vizinho, peso in lista[v_atual]:
            if distancias[v_atual] + peso >= distancias[vizinho]:
                continue
            distancias[vizinho] = distancias[v_atual] + peso
            heappush(fila_prioridade, (distancias[vizinho], vizinho))
            
    return distancias

n, m = map(int, input().split())
ilhas = [[] for i in range(n)]

for i in range(m):
    u, v, p = map(int, input().split())
    ilhas[u-1].append((v-1, p))
    ilhas[v-1].append((u-1, p))
    
s = int(input())

distancias = dijkistra(s-1, ilhas)
del distancias[s-1]
print(max(distancias) - min(distancias))

File: beecrowd/c++/grafos/ex2784.cpp
#include <bits/stdc++.h>
using namespace std;

typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;

int dij(int origem, int numVertices, vector<vii>lista){
    const int INF = 1e9;
    priority_queue<ii, vector<ii>, greater<ii>> fila_prioridade;
    fila_prioridade.push({0, origem});
    vi distancias(numVertices, INF);
    distancias[origem] = 0;

    while (!fila_prioridade.empty()){
        int dist = fila_prioridade.top().first;
        int v_atual = fila_prioridade.top().second;
        fila_prioridade.pop();

        if(dist > distancias[v_atual])
            continue;

        for (ii v : lista[v_atual]){
            int peso = v.second;
            int vz = v.first;
            if(distancias[v_atual] + peso >= distancias[vz])
                continue;
            distancias[vz] = distancias[v_atual] + peso;
            fila_prioridade.push({distancias[vz], vz});
        }
    }

    int menor, maior=0;
    if (distancias[0] !=0) 
        menor = distancias[0];
    else
        menor = distancias[1];
    for (int u=0; u<numVertices; u++){
        if(distancias[u] < menor && distancias[u] > 0)
            menor = distancias[u];
        if(distancias[u] > maior)
            maior = distancias[u];
    }

    return maior - menor;

}

int main(){
    int n, m, u, v, p, s;
    cin>>n>>m;

    vector<vii> lista(n, vii());
    for(int i=0; i<m; i++){
        cin>>u>>v>>p;
        lista[u-1].emplace_back(v-1, p);
        lista[v-1].emplace_back(u-1, p);
    }

    cin>>s;

    cout<<dij(s-1, n, lista)<<endl;

    return 0;
}